### PR TITLE
Provide a function to estimate decompressed size.

### PR DESCRIPTION
**Pull request by shakeelrao**
Provides a function to estimate the upper-bound of a compressed message.

Implements https://github.com/facebook/zstd/issues/1536.

### CODE REVIEW

**Comment by shakeelrao** [comment]
Sorry for all the failed CI runs. For some reason, the tests don't run to completion on my machine.

For example: `make shortest` outputs:

```
test : detect truncated compressed file 

truncatable-input.txt : 29.53%   ( 50000 =>  14766 bytes, truncatable-input.txt.zst) 
dd: unknown operand status
make[1]: *** [test-zstd] Error 1
make: *** [shortest] Error 2
```

Any suggestions?

File: lib/decompress/zstd_decompress.c [hunk 1]
```diff
@@ -483,10 +480,79 @@ size_t ZSTD_findFrameCompressedSize(const void *src, size_t srcSize)
...
+ *  `src` must point to the start of a ZSTD frame or a skippeable frame
+ *  `srcSize` must be at least as large as the frame contained
+ *  @return : maximum decompressed size of the compressed source
+ */
+unsigned long long ZSTD_decompressBound(const void* src, size_t srcSize)
+{
+    unsigned long long totalDstSize = 0;
+#if defined(ZSTD_LEGACY_SUPPORT) && (ZSTD_LEGACY_SUPPORT >= 1)
+    if (ZSTD_isLegacy(src, srcSize))
+        return ERROR(version_unsupported);
```
Comment (terrelln) [inline lib/decompress/zstd_decompress.c]: The error codes are `size_t`, but we are returning `unsigned long long`, so we can't return them.

When this function errors, return `ZSTD_CONTENTSIZE_ERROR`.

**Comment by terrelln** [comment]
To add a test with the frame content size isn't available, you can compress like this:

```c
ZSTD_CCtx* cctx = ZSTD_createCCtx();
CHECK_Z( ZSTD_CCtx_setParameter(cctx, ZSTD_c_contentSizeFlag, 0) );
CHECKPLUS(r, ZSTD_compress2(cctx,
                            compressedBuffer, compressedBufferSize,
                            CNBuffer, CNBuffSize),
                  cSize=r );
ZSTD_freeCCtx(cctx);
```

Look at the documentation for `ZSTD_c_contentSizeFlag` and `ZSTD_compress2()` to see whats going on here.

Just make sure that you do it after anything that relies on `compressedBuffer`, so either the last test, or if the next test also compresses into `compressedBuffer`.

**Comment by terrelln** [comment]
What are you running `make shortest` on? It works for me on Mac and Linux.

You can rely on the CI to run your tests. Facebook's CI will automatically run when you update the PR, but sometimes the queues get long, so you can set up CircleCI / Travis to run on your own repo if you like.

**Comment by shakeelrao** [comment]
Thanks for the detailed feedback, @terrelln! Really appreciate your patience on this :) Will start tackling some of the requested changes tonight after work. 

I am running `makeshortest` on Mac. For some reason the test fails when dealing with the `
truncatable-input.txt` test case.

File: lib/decompress/zstd_decompress.c [hunk 2]
```diff
@@ -434,12 +434,7 @@ static size_t ZSTD_decodeFrameHeader(ZSTD_DCtx* dctx, const void* src, size_t he
...
-/** ZSTD_findFrameCompressedSize() :
- *  compatible with legacy mode
- *  `src` must point to the start of a ZSTD frame, ZSTD legacy frame, or skippable frame
- *  `srcSize` must be at least as large as the frame contained
- *  @return : the compressed size of the frame starting at `src` */
-size_t ZSTD_findFrameCompressedSize(const void *src, size_t srcSize)
+static size_t ZSTD_findFrameCompressedSize_internal(const void *src, size_t srcSize, size_t* bound)
 {
 #if defined(ZSTD_LEGACY_SUPPORT) && (ZSTD_LEGACY_SUPPORT >= 1)
     if (ZSTD_isLegacy(src, srcSize))
```
[the remark this answers is not in the document]
Comment (shakeelrao) [reply lib/decompress/zstd_decompress.c]: Good idea! Do you think it's reasonable to return `ERROR(version_unsupported)` for now and add legacy support in another PR?

File: lib/decompress/zstd_decompress.c (hunk 1 above)
Comment (shakeelrao) [reply lib/decompress/zstd_decompress.c]: Should I change the function definition to `size_t ZSTD_decompressBound(const void* src, size_t srcSize)` instead? I'm a bit confused between the use of `size_t` and `unsigned long long` in this project.

File: lib/decompress/zstd_decompress.c [hunk 3]
```diff
@@ -433,60 +433,134 @@ static size_t ZSTD_decodeFrameHeader(ZSTD_DCtx* dctx, const void* src, size_t he
...
+                frameSizeInfo.compressedSize = ERROR(srcSize_wrong);
+                frameSizeInfo.decompressedBound = ERROR(srcSize_wrong);
+                return frameSizeInfo;
+            }
             ip += 4;
         }
 
-        return ip - ipstart;
+        frameSizeInfo.compressedSize = ip - ipstart;
+        frameSizeInfo.decompressedBound = (zfh.frameContentSize != ZSTD_CONTENTSIZE_UNKNOWN)
```
Comment (shakeelrao) [inline lib/decompress/zstd_decompress.c]: Looks like here we assign `size_t` to `unsigned long long`, but changed `decompressedBound` to `unsigned long long` means we won't be able to use the standard error codes. Do you have any suggestions?

**Comment by shakeelrao** [comment]
I think I implemented most of your suggest changes, including the documentation. Let me know if there are any more API level changes. I want to make sure the overall implementation is good before adding the final tests :)

File: lib/decompress/zstd_decompress.c [hunk 4]
```diff
@@ -433,60 +433,135 @@ static size_t ZSTD_decodeFrameHeader(ZSTD_DCtx* dctx, const void* src, size_t he
...
+/**
+ * Contains the compressed frame size and an upper-bound for the decompressed frame size.
+ * Note: before using `compressedSize` you must check for errors using ZSTD_isError().
+ *       similarly, before using `decompressedBound`, you must check for errors using:
+ *          `decompressedBound` != ZSTD_CONTENTSIZE_UNKNOWN
+ */
+typedef struct {
+    size_t compressedSize;
+    unsigned long long decompressedBound;
+} ZSTD_frameSizeInfo;
```
Comment (terrelln) [inline lib/decompress/zstd_decompress.c]: You'll save some code duplication, and make it harder miss setting a field if you use this helper function.

```c
static ZSTD_frameSizeInfo ZSTD_errorFrameSizeInfo(size_t err)
{
    ZSTD_frameSizeInfo frameSizeInfo;
    frameSizeInfo.compressedSize = err;
    frameSizeInfo.decompressedBound = ZSTD_CONTENTSIZE_UNKNOWN;
    return frameSizeInfo;
}
```

File: lib/decompress/zstd_decompress.c [hunk 5]
```diff
@@ -433,60 +433,135 @@ static size_t ZSTD_decodeFrameHeader(ZSTD_DCtx* dctx, const void* src, size_t he
...
+        size_t nbBlocks = 0;
         ZSTD_frameHeader zfh;
 
         /* Extract Frame Header */
-        {   size_t const ret = ZSTD_getFrameHeader(&zfh, src, srcSize);
-            if (ZSTD_isError(ret)) return ret;
-            RETURN_ERROR_IF(ret > 0, srcSize_wrong);
+        {
+            size_t const ret = ZSTD_getFrameHeader(&zfh, src, srcSize);
+            if (ZSTD_isError(ret)) {
```
Comment (terrelln) [inline lib/decompress/zstd_decompress.c]:
```c
if (ZSTD_isError(ret))
    return ZSTD_errorFrameSizeInfo(ret);
```

File: lib/decompress/zstd_decompress.c [hunk 6]
```diff
@@ -433,60 +433,135 @@ static size_t ZSTD_decodeFrameHeader(ZSTD_DCtx* dctx, const void* src, size_t he
...
+ *  `src` must point to the start of a ZSTD frame or a skippeable frame
+ *  `srcSize` must be at least as large as the frame contained
+ *  @return : the maximum decompressed size of the compressed source
+ */
+unsigned long long ZSTD_decompressBound(const void* src, size_t srcSize)
+{
+    unsigned long long bound = 0;
+    /* Iterate over each frame */
+    while (srcSize > 0) {
+        ZSTD_frameSizeInfo frameSizeInfo = ZSTD_findFrameSizeInfo(src, srcSize);
```
Comment (terrelln) [inline lib/decompress/zstd_decompress.c]: nit: Make every variable you can const. Every non-const variable should be mutated. It is a hint to the reader that this variable gets updated.

```c
    ZSTD_frameSizeInfo const frameSizeInfo = ...;
    size_t const compressedSize = ...;
    unsigned long long const decompressedBound = ...;
```

File: lib/zstd.h [hunk 7]
```diff
@@ -1103,6 +1103,21 @@ typedef enum {
...
+ *  currently incompatible with legacy mode
+ *  `src` must point to the start of a ZSTD frame or a skippeable frame
+ *  `srcSize` must be at least as large as the frame contained
+ *  @return : - the maximum decompressed size of the compressed source
+ *            - if an error occured: ZSTD_CONTENTSIZE_ERROR
+ *
+ *  note 1  : an error can occur if `src` points to a legacy frame or an invalid/incorrectly formatted frame.
+ *  note 2  : the bound is exact when Frame_Content_Size field is available in EVERY frame of `src`.
+ *  note 3  : when Frame_Content_Size isn't provided, the upper-bound for that frame is calculated by:
+ *              upper-bound = min(128 KB, Window_Size)
```
Comment (terrelln) [inline lib/zstd.h]: `upper-bound = # blocks * min(128 KB, Window_Size)`

File: lib/zstd.h [hunk 8]
```diff
@@ -1103,6 +1103,21 @@ typedef enum {
  *            however it does mean that all frame data must be present and valid. */
 ZSTDLIB_API unsigned long long ZSTD_findDecompressedSize(const void* src, size_t srcSize);
 
+/** ZSTD_decompressBound() :
+ *  currently incompatible with legacy mode
+ *  `src` must point to the start of a ZSTD frame or a skippeable frame
+ *  `srcSize` must be at least as large as the frame contained
```
Comment (terrelln) [inline lib/zstd.h]: This has the same requirements as `ZSTD_findDecompressedSize()`.

```
 *  `src` should point the start of a series of ZSTD encoded and/or skippable frames
 *  `srcSize` must be the _exact_ size of this series	 *  `srcSize` must be the _exact_ size of this series
 *       (i.e. there should be a frame boundary exactly at `srcSize` bytes after `src`)	 *       (i.e. there should be a frame boundary exactly at `srcSize` bytes after `src`)
```

File: lib/decompress/zstd_decompress.c (hunk 4 above)
Comment (shakeelrao) [reply lib/decompress/zstd_decompress.c]: Great idea! I'll add this.

File: lib/decompress/zstd_decompress.c (hunk 5 above)
Comment (shakeelrao) [reply lib/decompress/zstd_decompress.c]: Should I apply this style consistently throughout the code (i.e. if there's only one statement after the if, omit the brackets)?

File: lib/decompress/zstd_decompress.c (hunk 4 above)
Comment (shakeelrao) [reply lib/decompress/zstd_decompress.c]: What do you think about this?

```
static ZSTD_frameSizeInfo ZSTD_errorFrameSizeInfo(size_t err)
{
    return ZSTD_createFrameSizeInfo(err, ZSTD_CONTENTSIZE_ERROR);
}

static ZSTD_frameSizeInfo ZSTD_createFrameSizeInfo(size_t compressedSize, unsigned long long decompressedBound) {
    ZSTD_frameSizeInfo frameSizeInfo;
    frameSizeInfo.compressedSize = compressedSize;
    frameSizeInfo.decompressedBound = decompressedBound;
    return frameSizeInfo;
}
```

File: lib/zstd.h (hunk 7 above)
Comment (shakeelrao) [reply lib/zstd.h]: Nice catch!

File: tests/fuzzer.c [hunk 9]
```diff
@@ -901,6 +908,11 @@ static int basicUnitTests(U32 seed, double compressibility)
         if (r != CNBuffSize / 2) goto _output_error; }
     DISPLAYLEVEL(3, "OK \n");
 
+    DISPLAYLEVEL(3, "test%3i : get tight decompressed bound of multiple frames : ", testNb++);
+    {   unsigned long long const r = ZSTD_decompressBound(compressedBuffer, cSize);
+        if (r != CNBuffSize / 2) goto _output_error; }
+    DISPLAYLEVEL(3, "OK \n");
```
[the remark this answers is not in the document]
Comment (shakeelrao) [reply tests/fuzzer.c]: If I'm not mistaken, I think `compressedBuffer` contains a skippable [frame](https://github.com/facebook/zstd/pull/1543/files#diff-800cd24573dee9ac3476b9b256b07bf3L888).

[the remark this answers is not in the document]
Comment (shakeelrao) [reply tests/fuzzer.c]: Thanks! I believe this test case has a skippable frame [already](https://github.com/facebook/zstd/pull/1543/files#diff-800cd24573dee9ac3476b9b256b07bf3R916).

File: lib/decompress/zstd_decompress.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (shakeelrao) [reply lib/decompress/zstd_decompress.c]: Do you think it's reasonable to return an error for now and implement legacy support in a different PR?

[the remark this answers is not in the document]
Comment (terrelln) [reply lib/decompress/zstd_decompress.c]: Yeah, it should be done before the release, but you or I can do it in a separate PR.

**Comment by shakeelrao** [comment]
Thanks for all the help, @terrelln :) The final solution is much cleaner than my original commit. 

As a side note: I was under the impression that my commits would be squashed before merging into dev, so I was pretty liberal in my use of commits. Is this the standard for this project?

**Pull request merged**